Capture all throwables in ExitResult

regiskuckaertz opened the pull request:
Addresses https://github.com/scalaz/scalaz-zio/issues/96

~~[This test](https://github.com/scalaz/scalaz-zio/blob/master/core/jvm/src/test/scala/scalaz/zio/RTSSpec.scala#L369) never terminates when no errors are passed; investigating.~~

regiskuckaertz commented on core/jvm/src/main/scala/scalaz/zio/RTS.scala:
```diff
@@ -19,12 +19,13 @@ trait RTS {
-    case ExitResult.Terminated(t) => throw t
-    case ExitResult.Failed(e)     => throw Errors.UnhandledError(e)
+    case ExitResult.Completed(v)       => v
+    case ExitResult.Terminated(Nil)    => throw new Error("Fiber interrupted")
```
I don't like that, what about adding a new class in `Errors`?

jdegoes replied:
Definitely.

jdegoes commented on core/jvm/src/test/scala/scalaz/zio/PromiseSpec.scala:
```diff
@@ -73,7 +73,7 @@ class PromiseSpec extends Specification with RTS {
       for {
         p <- Promise.make[Exception, Int]
-        s <- p.done[Exception](ExitResult.Terminated(error))
+        s <- p.interrupt(error)
```
In the test cases, if the error is not important anymore, it can be eliminated.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/Fiber.scala:
```diff
@@ -31,12 +31,12 @@ trait Fiber[E, A] { self =>
    * successfully interrupted or has produced its result.
    */
-  def interrupt[E2](t: Throwable): IO[E2, Unit]
+  def interrupt[E2](ts: Throwable*): IO[E2, Unit]
```
Can we do two variants:

```scala
def interrupt: IO[E2, Unit] = interrupt0(Nil)

def interrupt(t: Throwable, ts: Throwable*): IO[E2, Unit] = interrupt0(t :: ts.toList)

protected def interrupt0[E2](ts: List[Throwable]): IO[E2, Unit]
```

This will allow a cleaner syntax for the case of no concrete cause.

regiskuckaertz replied:
Oh yeah, I didn't think about that 👌

jdegoes commented on core/jvm/src/main/scala/scalaz/zio/RTS.scala:
```diff
@@ -19,12 +19,13 @@ trait RTS {
-    case ExitResult.Failed(e)     => throw Errors.UnhandledError(e)
+    case ExitResult.Completed(v)       => v
+    case ExitResult.Terminated(Nil)    => throw new Error("Fiber interrupted")
+    case ExitResult.Terminated(t :: _) => throw t
```
Now `unsafeRun` should return the "first" error. So whether this is correct or not depends on the order in which the `Throwable` are stored. Most likely they would be stored in reverse order (first exception is at the end of the list), in which case this code is not correct.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/IO.scala:
```diff
@@ -709,7 +708,7 @@ object IO {
    * Terminates the fiber executing this action, running all finalizers.
    */
-  final def terminate[E, A](t: Throwable): IO[E, A] = new Terminate(t)
+  final def terminate[E, A](t: Throwable*): IO[E, A] = new Terminate(t.toList)
```
We should do the no-arg, one arg variant to make this simpler to use.

jdegoes commented on core/jvm/src/main/scala/scalaz/zio/RTS.scala:
```diff
@@ -633,7 +628,7 @@ private object RTS {
               this.noInterrupt += 1
 
-              curIo = IO.terminate[E, Any](die.get)
+              curIo = IO.terminate[E, Any](die.getOrElse(Nil): _*)
```
It should probably be left as `die.get` because if it's empty, there's a bug in the code and we'd like to know about it. The only reason `fold` et al aren't used is the extra allocation overhead.

jdegoes commented on core/jvm/src/main/scala/scalaz/zio/RTS.scala:
```diff
@@ -936,30 +931,31 @@ private object RTS {
-        case Executing(t0, joiners, killers) =>
-          if (!status.compareAndSet(oldStatus, Executing(t0.orElse(Some(t)), joiners, k :: killers))) kill0(t, k)
+        case Executing(ts0, joiners, killers) =>
+          if (!status.compareAndSet(oldStatus, Executing(Some(ts0.getOrElse(Nil) ++ ts), joiners, k :: killers)))
```
Note the ordering matters, though: if we want the first errors that killed it to be at the end of the list, then this would be correct.

jdegoes commented on core/jvm/src/main/scala/scalaz/zio/RTS.scala:
```diff
@@ -974,9 +970,9 @@ private object RTS {
 
         case s @ AsyncRegion(_, _, _, _, _, _) =>
-          val newStatus = s.copy(error = s.error.orElse(Some(t)), killers = k :: s.killers)
+          val newStatus = s.copy(errors = Some(s.errors.getOrElse(Nil) ++ ts), killers = k :: s.killers)
```
Ditto on ordering.

jdegoes commented on core/jvm/src/main/scala/scalaz/zio/RTS.scala:
```diff
@@ -1015,22 +1011,22 @@ private object RTS {
   }
   object FiberStatus {
-    final case class Executing[E, A](error: Option[Throwable],
+    final case class Executing[E, A](errors: Option[List[Throwable]],
```
Looks good: `None` means it's not been killed; `Some` means it has been killed, possibly with some errors.

jdegoes commented on core/jvm/src/test/scala/scalaz/zio/RTSSpec.scala:
```diff
@@ -283,7 +283,7 @@ class RTSSpec(implicit ee: ExecutionEnv) extends Specification with AroundTimeou
             .fork
       _ <- (ref.read <* IO.sleep[Void](1.millisecond)).doUntil(_.contains("start 1"))
-      _ <- f.interrupt(new RuntimeException("cancel"))
+      _ <- f.interrupt()
```
Change to `f.interrupt` here and elsewhere.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/ExitResult.scala:
```diff
@@ -28,14 +28,14 @@ sealed trait ExitResult[E, A] { self =>
-  final case class Terminated[E, A](error: Throwable) extends ExitResult[E, A]
+  final case class Completed[E, A](value: A)                extends ExitResult[E, A]
+  final case class Failed[E, A](error: E)                   extends ExitResult[E, A]
+  final case class Terminated[E, A](error: List[Throwable]) extends ExitResult[E, A]
```
We should define in scaladoc what the ordering is for the errors.

jdegoes commented on core/jvm/src/main/scala/scalaz/zio/RTS.scala:
```diff
@@ -238,11 +237,7 @@ private object RTS {
-        // context's unhandled exception handler:
-        _.foldRight(IO.unit[Void])((t, io) => io *> unhandled(t))
-      )
+      errors.flatMap(unhandled)
```
At this point, the function probably doesn't need to exist since it's far longer than it's definition.

regiskuckaertz commented:
@jdegoes Thank you 😃 At the moment no, the error ordering I knew I had to double-check. Will add tests to prove they appear in original order and amend the code accordingly.

jdegoes commented:
@regiskuckaertz This ready for another review?

regiskuckaertz commented:
@jdegoes not yet, sorry. Have friends over for a few days 😅 will try to squeeze some time tomorrow morning

regiskuckaertz commented:
haha, forgot to add the test 😅and it breaks. I'm struggling to find a way to actually accumulate errors, every path I explore automatically yields an `IO.Terminate` as soon as the first exception occurs.

regiskuckaertz commented:
I think accumulation could happen quite simply by passing a different handler to a forked fiber:
```
case IO.Tags.Fork =>
  val io = curIo.asInstanceOf[IO.Fork[_, E, Any]]
  val optHandler = io.handler
  val handler = combineHandlers(collectErrors, if (optHandler eq None) unhandled else optHandler.get)
  val value: FiberContext[_, Any] = fork(io.value, handler)

def collectErrors: Canceler = ts => {
  val oldStatus = status.get

  oldStatus match {
    case s @ AsyncRegion(ts2, _, _, _, _, _) =>
      if (!status.compareAndSet(oldStatus, s.copy(errors = ts2 ++ ts))) collectErrors(ts)
    case s @ Executing(ts2, _, _) =>
      if (!status.compareAndSet(oldStatus, s.copy(errors = ts2 ++ ts))) collectErrors(ts)
    case _ =>
  }
}
```

what do you think @jdegoes ?

jdegoes commented:
@regiskuckaertz Forking should not accumulate errors. Or rather, failure of a forked fiber must report its errors independently.

Has `terminate` been changed to accept multiple errors?

The test that should be added here goes something along these lines:

1. Fail with `e: E`
2. Ensure with something that throws some `Throwable`

Now you have 2 failures. If you fork this fiber and supervise the failures, you should get `UnhandledError(e)` followed by the `Throwable`. You can nest these things to increase the errors, e.g. have two finalizers that each throw errors.

regiskuckaertz commented:
Yes, I can see that now. My intuition is that if the finalizers run successfully, the action should yield an `ExitResult.Failed(UnhandledError(error))`; otherwise an `ExitResult.Terminated(UnhandledError(error) :: ts)` where `ts` may be `Nil`. If yes then the finalization should become a `IO[Void, Option[List[Throwable]]]`. Is my intuition correct?

Re- ordering: yes, I will amend the scaladoc to make it clearer.

regiskuckaertz commented:
... or, everything I just said is garbage and instead:
- `IO.supervise` and `IO.supervised` don't need a `Throwable` argument anymore
- all this error accumulation should only happen in a supervising fiber

jdegoes commented:
@regiskuckaertz It looks like you asked and answered your own questions, but let me know if I need to address anything, and also when you're ready for me to take another look.

Thanks for all your great work on this!

regiskuckaertz commented:
Nope, thanks for your patience on this! It took me a while to figure out that I got supervision completely wrong. Will revert these changes and go down that line instead 👍

regiskuckaertz commented:
@jdegoes I may have found the source of my confusion. When I run this:
```scala
unsafeRun(for {
  f <- IO
    .fail[Throwable, Unit](ExampleError)
    .ensuring(IO.sync(throw InterruptCause1))
    .ensuring(IO.sync(throw InterruptCause2))
    .ensuring(IO.sync(throw InterruptCause3))
    .fork0(ts => IO.sync { println("hey: " + ts) })
   _ <- f.join
} yield ())
```
I can see the following output:
```
hey: List(java.lang.Exception: Oh noes 1!)
hey: List(java.lang.Exception: Oh noes 2!)
hey: List(java.lang.Exception: Oh noes 3!)
hey: List(java.lang.Exception: Oh noes 1!, java.lang.Exception: Oh noes 2!, java.lang.Exception: Oh noes 3!)
hey: List(scalaz.zio.Errors$UnhandledError: An error was not handled by a fiber: java.lang.Exception: Oh noes!)
```
In the `Terminate` and `Fail` cases, there is a call to the supervisor when the bottom of the stack is reached. I updated it as follows, e.g. in `Fail` from
```scala
rts.submit(rts.unsafeRun(unhandled(Errors.UnhandledError(error) :: Nil)))
```
to
```scala
if (supervising > 0) {
  rts.submit(rts.unsafeRun(unhandled(Errors.UnhandledError(error) :: Nil)))
}
```

Now I get the result I expect, i.e. only the fourth line is printed out, but I'm not 100% sure this is correct.

regiskuckaertz commented:
Hmm, for some reason with that change some thread seem not to be cleaned up as my CPU is running on high even after the task finished 🤔

regiskuckaertz commented:
I have a feeling that this is either the correct solution, or that the line
```scala
val finalization = finalizer.flatMap(unhandled)
```
should be replaced with something that passes the list of throwables (call it `ts`) back to the branch where we've reached the bottom of the stack. If that list is empty, then `Errors.UnhandledError(error) :: ts` and `causes ++ ts` does not change behaviour.

jdegoes commented:
@regiskuckaertz There is a contract in place: ZIO may never lose any errors under any circumstances. Supervision should have _zero_ effect on error handling. The only effect of supervision, as defined by the `supervise` combinator, is that fibers forked by the action will be terminated when the action itself concludes (either normally or abnormally). So the `supervise` machinery is really a distraction.

Now I sometimes call the unhandled error handler a supervisor, because it supervises a fiber and handles all the unhandled errors. But that's not really the same thing as the supervise combinator and I should figure out more consistent terminology.

Now let's take your example:

```scala
unsafeRun(for {
  f <- IO
    .fail[Throwable, Unit](ExampleError)
    .ensuring(IO.sync(throw InterruptCause1))
    .ensuring(IO.sync(throw InterruptCause2))
    .ensuring(IO.sync(throw InterruptCause3))
    .fork0(ts => IO.sync { println("hey: " + ts) })
   _ <- f.join
} yield ())
```

In the old model, each error would be reported to the unhandled error handler, as it occurred. Then separately, the failure of the `IO.fail` (the action being executed, excluding all the finalizers attached to it) would be evident when calling `unsafeRun`, because `unsafeRun` uses a `defaultUnhandler` that logs the exceptions to the console.

If our contract now is that we can guarantee the `ExitResult` contains all errors, including errors in finalizers, then this is powerful enough to change the model: we can now send all errors at once, at the end of the fiber's life, and not have to send them individually as they occur.

Now our model for `ExitResult` is not precise. We lose information on the error. I think in light of your changes we should modify this:

```scala
final case class Completed[E, A](value: A) extends ExitResult[E, A]
final case class Failed[E, A](error: E, defects: List[Throwable]) extends ExitResult[E, A]
final case class Terminated[E, A](defects: List[Throwable]) extends ExitResult[E, A]
```

This more precise representation allows us to capture the `E`. We could go further than that and stick the throwables into `E`, e.g. `ExitResult[Either[(E, List[Throwable]), List[Throwable]], A]`, but this feels clunky and not much benefit.

With this more precise representation, we would expect to see the fiber terminate with: 

```scala 
Failed(ExampleError, InterruptCause1 :: InterruptCause2 :: InterruptCause2 :: Nil)
```

This can then be passed to the unhandled error handler. However, not exactly in this form, because the unhandled error handler requires `List[Throwable]`. So we have to convert the `E` into a `Throwable` using `UnhandledError`—something I don't like about the current model, but it's necessary due to the hierarchical nature of the error model. I think we can clean it up but it's too much to do right now.

I will review your code and suggest things as I see them, but from my perspective now, it's quite possible the reporting of the same error more than once is an artifact of the evolution of this design, and that it's no longer necessary to avoid losing any errors in any circumstances.

regiskuckaertz commented:
@jdegoes Thanks, I appreciate you took the time to put all this in words 🙏 

> But that's not really the same thing as the supervise combinator and I should figure out more consistent terminology.

I think it would be helpful yes; I've been trying to reconcile these two ideas and wondered why the handler was called `Canceler` instead of `Supervisor`, but now I understand. Also, I now see why my last change is completely wrong 😅 I'll call the handler "canceler" until then.

> If our contract now is that we can guarantee the ExitResult contains all errors, including errors in finalizers, then this is powerful enough to change the model: we can now send all errors at once, at the end of the fiber's life, and not have to send them individually as they occur.

That was my line of reasoning but I just want to make this clear so my next commit will align with your idea:
1- `ExitResult.Failed` is updated as you describe above
2- in the `Fail` and `Terminate` branches, if there is a finalization process, errors are _not_ passed to the canceler but stored in the `FiberStatus`: they will be handed over to the canceler in the next loop iteration. Note that I think `IO.terminate` and `Fiber.interrupt` with no argument should not be considered defects. Inside the RTS, interruption for a reason should be advertised with a custom exception, e.g. in `Supervise` the children should be interrupted with the new `FiberInterrupted`.
3- in the `Fail` case, there is a slight complication in that the action may be inside an `Attempt`. We can either forget the defects and return an `Either[E, A]`, or fail by wrapping the `E` in `UnhandledError`. What do you think?

> I will review your code and suggest things as I see them, but from my perspective now, it's quite possible the reporting of the same error more than once is an artifact of the evolution of this design, and that it's no longer necessary to avoid losing any errors in any circumstances.

The changes above will take care of that, won't it? I guess my concern is that `PureCanceler` is a black box and users of `fork0` might have bad surprises using it.

jdegoes commented:
> I think it would be helpful yes; I've been trying to reconcile these two ideas and wondered why the handler was called Canceler instead of Supervisor, but now I understand. Also, I now see why my last change is completely wrong 😅 I'll call the handler "canceler" until then.

These are three separate things:

1. **Unhandled error handler**. These are functions of the form `List[Throwable] => IO[Void, Unit]` (formerly, `Throwable => IO[Void, Unit]`). Sometimes, confusingly, I have called these things "supervisors", because they "supervise" errors coming from fibers. The main fiber has one of these defined in `RTS` (`defaultUnhandler`), which can be overridden. Others are specified at the time of `fork`ing, or they are inherited from the parent if not specified.
2. **Canceler**. These are functions in the form `Throwable => Unit` or (`Throwable => IO[Void, Unit]` — pure canceler). These functions are called when an async action that is still in progress is interrupted. They are called in order to save work or to clean up resources because the result of the async computation is no longer needed. 
3. **Supervision**. By "supervision", here, I refer to the `Supervise` operation encoded in the `supervised` and `supervise` method of `IO` class and companion object, respectively. In this context, supervision means that as soon as the `io` exits, anything it forked will be interrupted. This type of supervision is useful to eliminate fiber leaks—if you spawned a bunch of fibers, and are wrapping up a computation and don't need them to complete, you can have them auto-interrupted by using this supervisor concept.

Does that make more sense?

> 2- in the Fail and Terminate branches, if there is a finalization process, errors are not passed to the canceler but stored in the FiberStatus

I actually think we should delete the `Throwable` from the canceler to eliminate this concern. If we didn't delete it, I'd say it's correct to pass it errors from Terminate or Interrupt (keep in mind, interruption is just a form of termination). But I'd say it's simpler just to delete the throwable and make a canceler `() => Unit` or `IO[Void, Unit]`. In fact, at this point, we can just make all cancelers pure: `IO[Void, Unit]` since we can make the overhead almost the same with a little optimization.

> : they will be handed over to the canceler in the next loop iteration. Note that I think IO.terminate and Fiber.interrupt with no argument should not be considered defects. 

All these errors must end up as part of the `ExitResult` of the fiber. Maybe we should soften the language and change from "defects" to "causes".

> Inside the RTS, interruption for a reason should be advertised with a custom exception, e.g. in Supervise the children should be interrupted with the new FiberInterrupted.

Yes, possibly for `Supervise`, we should interrupt the children with a well-defined cause so developers can figure out why their fibers are dying (if they don't know how `supervise` works). 

See also: #124, which will generalize supervisor so we can make the current one implemented in `IO`, not in `RTS`, by using the more generic operation.

> 3- in the Fail case, there is a slight complication in that the action may be inside an Attempt. 

We only need to worry about this when there are no exception handlers. In this case, the stack will be unwound and be empty, and there will be nothing left, so fiber will be terminated and the error should be reported to the uncaught error handler.

> The changes above will take care of that, won't it? I guess my concern is that PureCanceler is a black box and users of fork0 might have bad surprises using it.

I'm not sure what you're talking about here. I think the use of the `PureCanceler` type synonym is confusing because the thing passed to `fork0` is not a pure canceler, or any type of canceler, it's an uncaught error handler.

regiskuckaertz commented:
> Does that make more sense?
Yes, this clear things up _a lot_ in my mind. I conflated canceler and error handler. Will update as follows if you don't mind:
```scala
type Canceler = Infallible[Unit]
type ErrorHandler = List[Throwable] => Infallible[Unit]
```
Indeed it is even mandatory to remove the `Throwable` from `Canceler`, because otherwise it might never get called (in the case where `ts == Nil` in `kill0`). I'll leave the unification of `PureCanceler` and `Canceler` to a follow-up PR.
> All these errors must end up as part of the ExitResult of the fiber.
👍 
> Yes, possibly for Supervise, we should interrupt the children with a well-defined cause
👍 
> We only need to worry about this when there are no exception handlers.
👍 
> I'm not sure what you're talking about here.
That's indeed because I've mistaken the two things to be the same all along. Have now fixed that, again thanks a lot for clarifying.

regiskuckaertz commented:
@jdegoes this is now ready for a thorough review 🎉

regiskuckaertz commented on core/jvm/src/test/scala/scalaz/zio/PromiseSpec.scala:
```diff
@@ -52,7 +52,7 @@ class PromiseSpec extends Specification with RTS {
       for {
         p <- Promise.make[String, Int]
-        s <- p.done[String](ExitResult.Failed("error in e4"))
+        s <- p.done[String](ExitResult.Failed("error in e4", Nil))
```
Something to think when reviewing the design of `ExitResult` is that errors are lost when the action fails within `attempt`, as only the `E` is produced in the value channel.

regiskuckaertz replied:
I haven't touched `attempt`, it would require changing it to something like `IO[Nothing, Either[(E, List[Throwable]), A]]` 😱

regiskuckaertz commented:
Still need to fix interruption

regiskuckaertz commented:
hmm, the test runs fine locally, there's a race condition somewhere 🤔

regiskuckaertz commented:
@jdegoes I'm at a loss to explain why this test breaks in 2.11:
```scala
(for {
      f1 <- IO.never[Void, Unit].fork
      _  <- f1.interrupt[Void](InterruptCause1, InterruptCause2)
      _  <- f1.join
} yield ()).run
```
produces `List(InterruptCause1, InterruptCause2, InterruptCause1, InterruptCause2)`. Could it be somehow related to #93 ?

regiskuckaertz commented:
@jdegoes I _think_ I identified the reason why it broke: 
- in `kill0`, the `killed` flag was set eagerly, regardless of whether the status of the fiber could be updated
- in the `shouldDie == Some(ts)` case, the fiber is terminated by passing the `ts`
- but since now defects are accumulated in the `Tags.Terminate` branch, this resulted in two concurrent processes updating the fiber status with the same value

To fix that, I:
- made sure the `killed` flag was only set when the status has been updated
- removed the list of failures in the call to `IO.terminate0` in the `shouldDie == Some(ts)` branch

Do you think this is valid?

jdegoes commented on core/jvm/src/main/scala/scalaz/zio/RTS.scala:
```diff
@@ -424,14 +412,21 @@ private object RTS {
                         curIo = null
-                        result = ExitResult.Failed(error)
+
+                        val ts: List[Throwable] = status.get match {
```
We could clean this up by adding a method to status called `causes`, which extracts out `List[Throwable]` from the interrupt causes.

jdegoes commented on core/jvm/src/main/scala/scalaz/zio/RTS.scala:
```diff
@@ -424,14 +412,21 @@ private object RTS {
 
                         // Report the uncaught error to the supervisor:
-                        rts.submit(rts.unsafeRun(unhandled(Errors.UnhandledError(error))))
+                        rts.submit(rts.unsafeRun(unhandled(Errors.UnhandledError(error) :: ts)))
```
I am now thinking this is not a good place to report the errors. We could consolidate the logic by putting it into `FiberContext.done`, and ensure that no matter how the fiber terminate, as soon as it's done, all the errors get reported in the same uniform way.

jdegoes commented on core/jvm/src/main/scala/scalaz/zio/RTS.scala:
```diff
@@ -554,25 +549,31 @@ private object RTS {
                       // No finalizers, simply produce error:
                       curIo = null
-                      result = ExitResult.Terminated(cause)
+                      val allCauses = status.get match {
```
Ditto for this, I think we push this to `causes` method on the status so the code is just `io.causes ++ status.get.causes`.

jdegoes commented on core/jvm/src/main/scala/scalaz/zio/RTS.scala:
```diff
@@ -554,25 +549,31 @@ private object RTS {
 
                       // Report the termination cause to the supervisor:
-                      rts.submit(rts.unsafeRun(unhandled(cause)))
+                      rts.submit(rts.unsafeRun(unhandled(allCauses)))
```
Ditto for this. It's too easy to miss calling this or to call it in a different way. Best to handle it uniformly.

jdegoes commented on core/jvm/src/main/scala/scalaz/zio/RTS.scala:
```diff
@@ -611,7 +612,7 @@ private object RTS {
               this.noInterrupt += 1
 
-              curIo = IO.terminate[E, Any](die.get)
+              curIo = IO.terminate0[E, Any](Nil)
```
This code is definitely wrong. I think it must be `IO.terminate0(status.get.causes)`.

regiskuckaertz replied:
I made that change because in `kill0` the interruptions are added to `status.errors`, so when this action is evaluated by RTS, the errors will already be there and picked up in the `Terminate` branch. Or do you prefer that I don't add the errors in `kill0` and instead revert `shouldDie` to return an optional list of throwables? As far as I can see, the code above will only be executed after `kill0` has been called.

regiskuckaertz replied:
Here's the [diff](https://github.com/scalaz/scalaz-zio/pull/107/commits/e954f3b0f3fe31dfe3498f1e5317bf77254307fd) for that particular change. Before that, I had the same exceptions reported twice in my test (only on Travis).

jdegoes commented on core/jvm/src/main/scala/scalaz/zio/RTS.scala:
```diff
@@ -635,14 +636,33 @@ private object RTS {
     }
 
+    private final def accumFailures: List[Throwable] => IO[Void, Unit] = {
+      case Nil => IO.now[Void, Unit](())
```
You can use `IO.unit` to avoid allocating something here.

regiskuckaertz replied:
Thanks!

jdegoes commented on core/jvm/src/main/scala/scalaz/zio/RTS.scala:
```diff
@@ -883,8 +903,8 @@ private object RTS {
     @inline
-    final def shouldDie: Option[Throwable] =
-      if (!killed || noInterrupt > 0) None else status.get.error
+    final def shouldDie: Option[List[Throwable]] =
```
Since you are not calling `get` on this, may as well just return a boolean from `shouldDie`.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/Errors.scala:
```diff
@@ -12,4 +12,7 @@ object Errors {
       extends Exception("An error was not handled by a fiber: " + error.toString())
+
+  final object InterruptedFiber
+      extends Exception("The fiber was interrupted by a user-defined action or its supervisor")
```
We don't use this anymore, do we?

jdegoes commented on core/shared/src/main/scala/scalaz/zio/ExitResult.scala:
```diff
@@ -10,32 +10,36 @@ sealed trait ExitResult[E, A] { self =>
-  final case class Failed[E, A](error: E)             extends ExitResult[E, A]
-  final case class Terminated[E, A](error: Throwable) extends ExitResult[E, A]
+  final case class Completed[E, A](value: A)                       extends ExitResult[E, A]
+  final case class Failed[E, A](error: E, causes: List[Throwable]) extends ExitResult[E, A]
```
These are not so much `causes` as a combination of finalizer defects (exceptions thrown from finalizers) and maybe interruption causes. This conflation makes me want to clean up the model and state:

 1. If `ExitResult.Failed`, then the fiber failed because it did not handle an `E` error, and in the process of terminating the fiber, finalizers were ran which may have resulted in some runtime exceptions. No interruption causes are reported because even if the fiber was interrupted at the same time it failed, ultimately the cause of its failure is the unhandled `E` error.
 2. If `ExitResult.Terminated`, then the fiber failed either because of interruption, or because of a defect, and any information on the failure is reported in `causes`.

It would be possible to further separate (2), but I'm not sure it's worth it.

regiskuckaertz replied:
I admit I like the current `Failed` better. It's opinionated: the developer is accountable for exceptions, otherwise the whole program breaks down; there's no negotiation. If a finalizer throws, it may be a small thing, or it can have catastrophic consequences—I think the library should make the worst assumption.

And what to make of that asymmetry: if an `A` is produced in the value channel but finalizers throw exceptions, the action yields a `Terminated`.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/Fiber.scala:
```diff
@@ -31,12 +31,25 @@ trait Fiber[E, A] { self =>
+  /**
+   * Interrupts the fiber with a list of error(s).
+   */
+  protected def interrupt0[E2](ts: List[Throwable]): IO[E2, Unit]
```
We could probably make this public.

regiskuckaertz commented:
As a user of zio, `IO.fail(e, ts)` doesn't feel right to me. If I exercise due diligence and design errors properly, having to deal with an additional list of exceptions feels antithetical to the design of `IO[E, A]`. I'm happy if my program breaks sooner because of an oversight in my code—then I can take immediate action to fix it.

The distinction between interruptions (an interruption can be made on purpose) and exceptions feels logical to me, but then when it is time to report those to the handler, this line feels somewhat awkward:
```scala
rts.submit(rts.evaluate(unhandled(UnhandleError(error, causes) :: interruptions)))
```

regiskuckaertz commented:
@jdegoes here's a summary of the latest changes for your reviewing:

- `FiberStatus` now manages an list of defects and an optional list of interruption causes: exceptions thrown will accumulate in the former, whereas the latter is only touched when the fiber is killed
- moved error reporting in `done` for consistency
- `collectDefect` returns an `Option[List[Throwable]]`, where the list is guaranteed to be non-empty (I'm not sure this is necessary, the `None` case will end up as `Nil` in the `ExitResult`)
- added `TerminatedFiber` when the fiber has been terminated with no reason (`InterruptedFiber` is used only when a parent fiber interrupts its children if supervising them)
- `shouldDie` returns a boolean (termination causes have already been accumulated in the fiber status)

The build is failing in Travis for some private method in `Fiber` that does not exist; I've removed all my `target` folders but am unable to reproduce locally.

jozic commented on core/shared/src/main/scala/scalaz/zio/Errors.scala:
```diff
@@ -10,6 +10,10 @@ object Errors {
+
+  final object TerminatedFiber extends Exception("The fiber was terminated either by a defect or an interruption")
+
+  final object InterruptedFiber extends Exception("The fiber was interrupted either by its superviser")
```
should `either` be dropped here?

regiskuckaertz replied:
oh, good catch 👌

jdegoes commented:
@regiskuckaertz If you can fix the merge conflicts, I'll take what I hope will be a final review!

regiskuckaertz commented:
@jdegoes sure thing!

regiskuckaertz commented:
Hi @jdegoes do you still have appetite for this one? 😁

jdegoes commented:
@regiskuckaertz Yes, sorry for the delay. Will try to do it this weekend!

jdegoes commented on core/shared/src/main/scala/scalaz/zio/IO.scala:
```diff
@@ -715,7 +712,17 @@ object IO {
    * Terminates the fiber executing this action, running all finalizers.
    */
-  final def terminate(t: Throwable): IO[Nothing, Nothing] = new Terminate(t)
+  final def terminate: IO[Nothing, Nothing] = terminate0(Nil)
```
Looks good.

jdegoes commented:
@regiskuckaertz This ready for a merge? We can create tickets for outstanding issues like dropping errors.

regiskuckaertz commented:
@jdegoes Yes 👍 I was about to ask precisely that.

jdegoes commented:
@regiskuckaertz Thanks for all your hard work on this!!!

tonymorris commented on core/jvm/src/main/scala/scalaz/zio/RTS.scala:
```diff
@@ -286,19 +281,22 @@ private object RTS {
+        case (None, ts @ Some(_))   => ts
+        case (None, None)           => None
+      }
+
```
This is a re-implementation of one of the `Semigroup` instances for `Option[A]`.

jdegoes replied:
It's true, but now Scalaz depends on ZIO (which means, at the very least, no more impure code in Scalaz), so the reimplementation is necessary.
